### PR TITLE
LRGB: Long Range Graph Benchmark datasets

### PR DESCRIPTION
**Pull request by vijaydwivedi75**
This PR adds the LRGB datasets from the paper [Long Range Graph Benchmark](https://openreview.net/pdf?id=in7XC5RcjEn). The original dataset source is [in this repo](http://github.com/vijaydwivedi75/lrgb).

The Long Range Graph Benchmark (LRGB) is a collection of 5 graph learning datasets that arguably require long-range reasoning to achieve strong performance in a given task. The 5 datasets in this benchmark can be used to prototype new models that can capture long range dependencies in graphs.

Dataset | Domain | Task
-- | -- | --
PascalVOC-SP | Computer Vision | Node Classification
COCO-SP | Computer Vision | Node Classification
PCQM-Contact | Quantum Chemistry | Link Prediction
Peptides-func | Chemistry | Graph Classification
Peptides-struct | Chemistry | Graph Regression

The `torch_geometric.datasets.LRGBDataset` can be used to access any of the 5 datasets in the benchmark.

### CODE REVIEW

**Comment by EdisonLeeeee** [comment]
Thanks for adding these! Left some initial comments.

File: torch_geometric/datasets/lrgb.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+
+    def download(self):
+        shutil.rmtree(self.raw_dir)
+        path = download_url(self.urls[self.name], self.root)
+        extract_zip(path, self.root)
+        os.rename(osp.join(self.root, self.dwnld_file_name[self.name]),
+                  self.raw_dir)
+        os.unlink(path)
+
+    def process(self):
```
Comment (EdisonLeeeee) [inline torch_geometric/datasets/lrgb.py:158]: Since `process_sp`  and `process_peptides` work in a similar way, can we unify them into one single processing function?

File: torch_geometric/datasets/lrgb.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+                x = graph[0].to(torch.float)
+                edge_attr = graph[1].to(torch.float)
+                edge_index = graph[2]
+                y = torch.LongTensor(graph[3])
+
+                if self.name == 'coco-sp':
+                    # Label remapping for coco-sp.
+                    # See self.label_remap_coco() func
+                    label_map = self.label_remap_coco()
```
Comment (EdisonLeeeee) [inline torch_geometric/datasets/lrgb.py:198]: This can be simplified with `sklearning.preprocessing.LabelEncoder`, e.g., 
```python
from sklearn.preprocessing import LabelEncoder
y = LabelEncoder().fit_transform(graph[3])
y = torch.LongTensor(y)
```

File: torch_geometric/datasets/lrgb.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+            value, indicating whether the data object should be included in the
+            final dataset. (default: :obj:`None`)
+
+    Stats:
+        .. list-table::
+            :widths: 15 15 10 10 10 10
+            :header-rows: 1
+
+            * - Name
+              - Task
```
Comment (EdisonLeeeee) [inline torch_geometric/datasets/lrgb.py:55]: An additional column `Task` seems not supported in the `CheatSheet` part

**Comment by vijaydwivedi75** [comment]
Thanks @EdisonLeeeee for your comments! 
I have included the following revisions in an updated commit.
- Merged process steps for the superpixels and Peptides datasets
- Removed 'Task' column from the Stats
- Simplified contiguous label through LabelEncoder.

File: torch_geometric/datasets/lrgb.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+                              'rb') as f:
+                        graphs = torch.load(f)
+
+                indices = range(len(graphs))
+
+                pbar = tqdm(total=len(indices))
+                pbar.set_description(f'Processing {split} dataset')
+
+                data_list = []
+                for idx in indices:
```
Comment (EdisonLeeeee) [inline torch_geometric/datasets/lrgb.py:179]:
```suggestion
                for graph in tqdm(graphs, desc=f'Processing {split} dataset'):
```

File: torch_geometric/datasets/lrgb.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+                    where,
+                    num_labeled_edges are negative edges and link pred labels,
+                    https://github.com/vijaydwivedi75/lrgb/blob/main/graphgps/loader/dataset/pcqm4mv2_contact.py#L192
+                """
+                x = graph[0]
+                edge_attr = graph[1]
+                edge_index = graph[2]
+                edge_index_labeled = graph[3]
+                edge_label = graph[4]
+                y = None
```
Comment (EdisonLeeeee) [inline torch_geometric/datasets/lrgb.py:275]:
```suggestion
```

File: torch_geometric/datasets/lrgb.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+                data = Data(x=x, edge_index=edge_index, edge_attr=edge_attr,
+                            edge_index_labeled=edge_index_labeled,
+                            edge_label=edge_label, y=y)
```
Comment (EdisonLeeeee) [inline torch_geometric/datasets/lrgb.py:279]:
```suggestion
                            edge_label=edge_label)
```

File: torch_geometric/datasets/lrgb.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+                            Shape of x : [num_nodes, 14]
+                            Shape of edge_attr : [num_edges, 2]
+                            Shape of edge_index : [2, num_edges]
+                            Shape of y : [num_nodes]
+                        """
+                        x = graph[0].to(torch.float)
+                        edge_attr = graph[1].to(torch.float)
+                        edge_index = graph[2]
+                        if self.name == 'coco-sp':
+                            y = label_map_coco.transform(graph[3])
```
Comment (EdisonLeeeee) [inline torch_geometric/datasets/lrgb.py:195]: I think we can directly use `LabelEncoder.fit_transform(graph[3])` here without the need of `original_label_idxs`, wdyt?

Comment (vijaydwivedi75) [reply torch_geometric/datasets/lrgb.py:195]: This is because one graph sample will only have a subset of the labels present in `original_label_idxs`, and hence would require to fit on the full label set.

File: torch_geometric/datasets/lrgb.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
+import os
+import os.path as osp
+import pickle
+import shutil
+from typing import Callable, List, Optional
+
+import torch
+from sklearn.preprocessing import LabelEncoder
```
[the remark this answers is not in the document]
Comment (vijaydwivedi75) [reply torch_geometric/datasets/lrgb.py:8]: If sklearn is not there by default and the `original_label_idxs` is required, should I revert back to using the `label_remap()` function that was in the first version?

File: torch_geometric/datasets/lrgb.py (hunk 7 above)
Comment (EdisonLeeeee) [reply torch_geometric/datasets/lrgb.py:195]: Oh, I see. You are right!

File: torch_geometric/datasets/lrgb.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (EdisonLeeeee) [reply torch_geometric/datasets/lrgb.py:8]: I have two suggestions:
+ Import `sklearn` internally as https://github.com/pyg-team/pytorch_geometric/blob/062556307bf6ebca3f0f185db8fd2dafd7362313/torch_geometric/nn/models/node2vec.py#L173
+ Implement `LabelEncoder` simply with PyTorch. FYI:
```python
label_encoder = torch.zeros(max(original_label_idxs))
label_encoder[original_label_idxs] = torch.arange(len(original_label_idxs))
...

y = label_encoder[torch.LongTensor[graphs[2]])
```

[the remark this answers is not in the document]
Comment (vijaydwivedi75) [reply torch_geometric/datasets/lrgb.py:8]: I feel we can skip the usage of sklearn as (i) its a one-time thing for `COCO-SP` and does not make much difference per my runs, and (ii) we require the label set `original_label_idxs`.

[the remark this answers is not in the document]
Comment (EdisonLeeeee) [reply torch_geometric/datasets/lrgb.py:8]: Agreed. Not a strong preference for the use of sklearn.

**Comment by vijaydwivedi75** [comment]
Hi @EdisonLeeeee I have updated the code based on the feedbacks -thank you! 
It is ready from my side.

File: torch_geometric/datasets/lrgb.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+                        """
+                        x = graph[0]
+                        edge_attr = graph[1]
+                        edge_index = graph[2]
+                        y = graph[3]
+
+                    if self.name == 'coco-sp':
+                        # Label remapping for coco-sp.
+                        # See self.label_remap_coco() func
+                        label_map = self.label_remap_coco()
```
Comment (EdisonLeeeee) [inline torch_geometric/datasets/lrgb.py:202]: We are not necessarily to call this method and get `label_map` at each iteration, right?

Comment (vijaydwivedi75) [reply torch_geometric/datasets/lrgb.py]: Right -thanks! It can be called just once for coco. I updated the file.

File: CHANGELOG.md [hunk 10]
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ## [2.2.0] - 2022-MM-DD
 ### Added
+- Added `LRGBDataset` to include 5 datasets from the [Long Range Graph Benchmark](https://openreview.net/pdf?id=in7XC5RcjEn), [#5935](https://github.com/pyg-team/pytorch_geometric/pull/5935)
```
Comment (EdisonLeeeee) [inline CHANGELOG.md:8]:
```suggestion
- Added `LRGBDataset` to include 5 datasets from the [Long Range Graph Benchmark](https://openreview.net/pdf?id=in7XC5RcjEn) ([#5935](https://github.com/pyg-team/pytorch_geometric/pull/5935))
```

File: torch_geometric/datasets/lrgb.py [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+                    Shape of edge_label : [num_labeled_edges]
+
+                    where,
+                    num_labeled_edges are negative edges and link pred labels,
+                    https://github.com/vijaydwivedi75/lrgb/blob/main/graphgps/loader/dataset/pcqm4mv2_contact.py#L192
+                """
+                x = graph[0]
+                edge_attr = graph[1]
+                edge_index = graph[2]
+                edge_index_labeled = graph[3]
```
Comment (EdisonLeeeee) [inline torch_geometric/datasets/lrgb.py:264]: [naming] Use `edge_label_index` to be consistent with data split by `torch_geometric.transforms.RandomLinkSplit`.

**Review by EdisonLeeeee — APPROVED** [review]

**Comment by EdisonLeeeee** [comment]
A gentle ping @rusty1s :)

File: torch_geometric/datasets/lrgb.py (hunk 11 above)
Comment (vijaydwivedi75) [reply torch_geometric/datasets/lrgb.py]: Renamed `edge_index_labeled` to `edge_label_index`.

**Comment by vijaydwivedi75** [comment]
Done, added dataset info in docstring!

**Review by wsad1 — APPROVED** [review]

File: torch_geometric/datasets/lrgb.py [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+              - ~150.94
+              - ~307.30
+              - 10
+            * - Peptides-struct
+              - 15,535
+              - ~150.94
+              - ~307.30
+              - 11
+
+    """
```
Comment (rusty1s) [inline torch_geometric/datasets/lrgb.py:99]:
```suggestion
              - 11
    """
```

File: torch_geometric/datasets/lrgb.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+        'coco-sp': 'coco_superpixels_edge_wt_region_boundary',
+        'pcqm-contact': 'pcqmcontact',
+        'peptides-func': 'peptidesfunc',
+        'peptides-struct': 'peptidesstruct'
+    }
+
+    def __init__(self, root: str, name: str, split: str = "train",
+                 transform: Optional[Callable] = None,
+                 pre_transform: Optional[Callable] = None,
+                 pre_filter: Optional[Callable] = None):
```
Comment (rusty1s) [inline torch_geometric/datasets/lrgb.py:130]:
```suggestion
                 pre_filter: Optional[Callable] = None, ):
```

**Review by rusty1s — APPROVED** [review]
Thank you for adding <3
